controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -532,6 +545,62 @@ def test_create_volume_with_zero_size(self, storage_agent):
         self.assertEqual(self.context.code, grpc.StatusCode.OK)
         self.mediator.create_volume.assert_called_once_with(self.request.name, 1 * 1024 * 1024 * 1024, {}, "pool1")
 
+    def _prepare_idempotent_tests(self):
+        self.mediator.get_volume = Mock()
+        self.mediator.copy_to_existing_volume_from_source = Mock()
+        self.request.volume_content_source = self._get_source_snapshot("wwn1")
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent(self, storage_agent):
+        self._prepare_idempotent_tests()
+        storage_agent.return_value = self.storage_agent
+        snapshot_id = "wwn1"
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2", "a9k",
+                                                                                        copy_source_id=snapshot_id)
+
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.OK)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, snapshot_id)
+        self.mediator.copy_to_existing_volume_from_source.assert_not_called()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent_no_source(self, storage_agent):
+        self._prepare_idempotent_tests()
+        storage_agent.return_value = self.storage_agent
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2",
+                                                                                        "a9k")
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.ALREADY_EXISTS)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, '')
+        self.mediator.copy_to_existing_volume_from_source.assert_not_called()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent_source_not_requested_but_found(self, storage_agent):

```

Comment:
then why `from_source`?

Reply:
done

Reply:
1. `test_create_idempotent` should be `test_create_volume_idempotent` here.
2. now that `test_create_volume_from_source_idempotent_no_source` was changed to `test_create_volume_idempotent_with_source_volume_have_no_source`, IMO it is more ambiguous (request vs. existing volume) - perhaps `idempotent` is not a must, maybe just `test_create_volume_from_source_already_exists_without_source`?

Reply:
but its a specific test for idempotent

Reply:
1 is done